src/main/java/ch/njol/skript/expressions/ExprFurnaceEventValues.java

```diff
@@ -0,0 +1,156 @@
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.classes.Changer.ChangeMode;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Events;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.expressions.base.EventValueExpression;
+import ch.njol.skript.expressions.base.PropertyExpression;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.ExpressionType;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.util.Kleenean;
+import ch.njol.util.coll.CollectionUtils;
+import org.bukkit.Material;
+import org.bukkit.block.Block;
+import org.bukkit.event.Event;
+import org.bukkit.event.inventory.FurnaceBurnEvent;
+import org.bukkit.event.inventory.FurnaceExtractEvent;
+import org.bukkit.event.inventory.FurnaceSmeltEvent;
+import org.bukkit.event.inventory.FurnaceStartSmeltEvent;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.Nullable;
+
+
+@Name("Furnace Event Values")
+@Description({
+	"Represents the expressions you can use within furnace events to get items correlated to an event.",
+	"Only 'smelting item' can be changed."
+})
+@Examples({
+	"on furnace smelt:",
+		"\tbroadcast smelted item",
+		"\t# Or 'result'",
+	"on furnace extract:",
+		"\tbroadcast extracted item",
+	"on fuel burn:",
+		"\tbroadcast fuel burned",
+	"on smelting start:",
+		"\tbroadcast smelting item",
+		"\tclear smelting item"
+})
+@Events({"smelt", "fuel burn", "smelting start", "furnace extract"})
+@Since("INSERT VERSION")
+public class ExprFurnaceEventValues extends PropertyExpression<Block, ItemStack> {
+
+	enum FurnaceValues {
+		SMELTED("(smelted item|result [item])", FurnaceSmeltEvent.class, "Can only use 'smelted item' in a smelting event."),
+		EXTRACTED("extracted item[s]", FurnaceExtractEvent.class, "Can only use 'extracted item' in a furnace extract event."),
+		SMELTING("smelting item", FurnaceStartSmeltEvent.class, "Can only use 'smelting item' in a start smelting event"),
+		BURNED("fuel burned [item]", FurnaceBurnEvent.class, "Can only use 'fuel burned' in a fuel burning event.");
+
+		private String pattern, error;
+		private Class<? extends Event> clazz;
+
+		FurnaceValues(String pattern, Class<? extends Event> clazz, String error) {
+			this.pattern = "[the] " + pattern;
+			this.clazz = clazz;
+			this.error = error;
+		}
+
+	}
+
+	private static final FurnaceValues[] furnaceValues = FurnaceValues.values();
+
+	static {
+		int size = furnaceValues.length;
+		String[] patterns  = new String[size];
+		for (FurnaceValues value : furnaceValues) {
+			patterns[value.ordinal()] = value.pattern;
+		}
+
+		Skript.registerExpression(ExprFurnaceEventValues.class, ItemStack.class, ExpressionType.PROPERTY, patterns);
+	}
+
+	private FurnaceValues type;
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
```

Comment:
should have a newLine after matchedPattern

Reply:
idk wym by this